#include "detectmainwidget.h"
#include "ui_detectmainwidget.h"
#include <QtConcurrent>
#include "clientUtil.h"
#include "network.h"
#include <QPixmap>
#include <QDebug>
#include "maildialog.h"
#include "sensorreceive.h"

DetectMainWidget::DetectMainWidget(QStackedWidget *stack, QWidget *parent)
    : DetectCoreWidget(stack, parent)
    , ui(new Ui::DetectMainWidget)
{
    ui->setupUi(this);

    // ìŠ¤íŠ¸ë¦¼ ì •ë³´ ì„¤ì •
    cctvNames.emplace(0, "Fire");
    cctvNames.emplace(1, "Intrusion");
    cctvNames.emplace(2, "Strawberry");
    cctvNames.emplace(3, "Map");

    videoLabels[0] = ui->videoLabel1;
    videoLabels[1] = ui->videoLabel2;
    videoLabels[2] = ui->videoLabel3;
    videoLabels[3] = ui->videoLabel4;

    for (int i = 0; i < 4; ++i)
        videoLabels[i]->setScaledContents(true);

    // ë²„íŠ¼ ì—°ê²° ìœ ì§€
    connect(ui->btnFire,       &QToolButton::clicked, this, &DetectMainWidget::showFirePage);
    ui->btnFire->setMinimumSize(100, 100);
    ui->btnFire->setIcon(QIcon(":/icons/fire2.png"));
    ui->btnFire->setIconSize(QSize(75, 75));  // ì•„ì´ì½˜ í¬ê¸° ì¡°ì ˆ
    ui->btnFire->setToolButtonStyle(Qt::ToolButtonTextBesideIcon); // ì•„ì´ì½˜ ìœ„, í…ìŠ¤íŠ¸ ì•„ë˜

    connect(ui->btnGrowth,     &QToolButton::clicked, this, &DetectMainWidget::showGrowthPage);
    ui->btnGrowth->setMinimumSize(100, 100);
    ui->btnGrowth->setIcon(QIcon(":/icons/strawberry2.png"));
    ui->btnGrowth->setIconSize(QSize(75, 75));  // ì•„ì´ì½˜ í¬ê¸° ì¡°ì ˆ
    ui->btnGrowth->setToolButtonStyle(Qt::ToolButtonTextBesideIcon); // ì•„ì´ì½˜ ìœ„, í…ìŠ¤íŠ¸ ì•„ë˜

    connect(ui->btnIntrusion,  &QToolButton::clicked, this, &DetectMainWidget::showIntrusionPage);
    ui->btnIntrusion->setIcon(QIcon(":/icons/intrusion.png"));
    ui->btnIntrusion->setIconSize(QSize(75, 75));  // ì•„dì´ì½˜ í¬ê¸° ì¡°ì ˆ
    ui->btnIntrusion->setToolButtonStyle(Qt::ToolButtonTextBesideIcon); // ì•„ì´ì½˜ ìœ„, í…ìŠ¤íŠ¸ ì•„ë˜

    connect(ui->btnLog,        &QToolButton::clicked, this, &DetectMainWidget::showLogPage);
    ui->btnLog->setIcon(QIcon(":/icons/log.png"));
    ui->btnLog->setIconSize(QSize(75, 75));  // ì•„ì´ì½˜ í¬ê¸° ì¡°ì ˆ
    ui->btnLog->setToolButtonStyle(Qt::ToolButtonTextBesideIcon); // ì•„ì´ì½˜ ìœ„, í…ìŠ¤íŠ¸ ì•„ë˜

    connect(ui->btnMail, &QToolButton::clicked, this, &DetectMainWidget::showMailDialog);
    //ui->btnMail->setMinimumSize(100, 100); // ë‹¤ë¥¸ ë²„íŠ¼ê³¼ í†µì¼
    ui->btnMail->setIcon(QIcon(":/icons/mail2.png"));
    ui->btnMail->setIconSize(QSize(50, 50));

    connect(ui->btnCamera, &QToolButton::clicked, this, &DetectMainWidget::setupAllStreams);
    ui->btnCamera->setIcon(QIcon(":/icons/camera.png"));
    ui->btnCamera->setIconSize(QSize(50, 50));

    ui->btnSync->setIcon(QIcon(":/icons/car2.png"));
    ui->btnSync->setIconSize(QSize(50, 50));

    weatherManager = new QNetworkAccessManager(this);
    connect(weatherManager, &QNetworkAccessManager::finished, this, &DetectMainWidget::handleWeatherData);

    fetchWeather();  // ë‚ ì”¨ ê°€ì ¸ì˜¤ê¸° ìš”ì²­

    // ìŠµë„/ì˜¨ë„ í‘œì‹œ íƒ€ì´ë¨¸
    QTimer* sensorDisplayTimer = new QTimer(this);
    connect(sensorDisplayTimer, &QTimer::timeout, this, [=]() {
        float temp = sensorData.tempValue;
        float humid = sensorData.humidValue;

        ui->label_temhum->setText(
            QString("ğŸŒ¡ %1Â°C   ğŸ’§ %2%")
                .arg(temp, 0, 'f', 1)
                .arg(humid, 0, 'f', 1)
            );
    });
    sensorDisplayTimer->start(2000);  // 2ì´ˆë§ˆë‹¤ ì—…ë°ì´íŠ¸

}

DetectMainWidget::~DetectMainWidget()
{
    delete ui;
}

void DetectMainWidget::on_btnSync_clicked(){ sendFile("start", "CMD"); }

void DetectMainWidget::showFirePage()     { changePage(1); }
void DetectMainWidget::showGrowthPage()   { changePage(3); }
void DetectMainWidget::showIntrusionPage(){ changePage(2); }
void DetectMainWidget::showLogPage()   { changePage(4); }

void DetectMainWidget::setupStream(const char* cctvName, const int index){
    sendFile(cctvName, "IP");
    char buffer[1024];
    int n = SSL_read(sock_fd, buffer, sizeof(buffer) - 1);
    if (n > 0) {
        buffer[n] = '\0';
        QString ip = QString::fromUtf8(buffer);
        if(ip == "NO") {
            ipAddress[index].clear();
            return;
        }

        qDebug() << "[TEST]" << cctvName <<" ip:" << ip;
        ipAddress[index] = ip.toStdString();
        QString url = "rtsps://" + ip + ":8555/test";

        Streamers[index] = new VideoStreamHandler(index, url);

        connect(Streamers[index], &VideoStreamHandler::frameReady, this, [=](int idx, QImage img) {
            if (idx >= 0 && idx < 4 && videoLabels[idx]) {
                videoLabels[idx]->setPixmap(QPixmap::fromImage(img).scaled(videoLabels[idx]->size(), Qt::KeepAspectRatio, Qt::SmoothTransformation));
            }
        }, Qt::QueuedConnection);

        //Streamers[index]->initialize();  // ìë™ ì¬ì—°ê²° í¬í•¨ëœ ì´ˆê¸°í™”
    } else {
        qDebug() << "SSL_read failed or no data.";
        ipAddress[index].clear();
    }
}

void DetectMainWidget::setupAllStreams()
{
    qDebug() << "setupAllStreams() called";

    for(int i = 0; i < 4; i++){
        QTimer::singleShot(i * 1000, this, [=]() {
            if (ipAddress[i].empty()) {
                auto it = cctvNames.find(i);
                if (it != cctvNames.end()) {
                    setupStream(it->second.c_str(), i);
                }
            }
        });
    }
}

void DetectMainWidget::showMailDialog()
{
    MailDialog dialog(this);
    if (dialog.exec() == QDialog::Accepted) {
        QString email = dialog.getEmail();
        QString password = dialog.getPassword();
        qDebug() << "[MailDialog] Email:" << email << ", Password:" << password;
    }
}

// ë‚ ì”¨ ì •ë³´ ìš”ì²­
void DetectMainWidget::fetchWeather()
{
    QString apiKey = "ed464dec9a556ac76d138891022188b5";
    QString city = "Seoul";
    QString urlStr = QString("https://api.openweathermap.org/data/2.5/weather?q=%1&appid=%2&units=metric&lang=kr")
                         .arg(city).arg(apiKey);

    QNetworkReply* reply = weatherManager->get(QNetworkRequest(QUrl(urlStr)));
    connect(reply, &QNetworkReply::finished, this, &DetectMainWidget::handleWeatherData);
}

// ë‚ ì”¨ ì•„ì´ì½˜ ìš”ì²­
void DetectMainWidget::fetchWeatherIcon(const QString& iconCode)
{
    QString iconUrlStr = QString("https://openweathermap.org/img/wn/%1@2x.png").arg(iconCode);
    QNetworkReply* iconReply = weatherManager->get(QNetworkRequest(QUrl(iconUrlStr)));
    connect(iconReply, &QNetworkReply::finished, this, &DetectMainWidget::handleWeatherIcon);
}

void DetectMainWidget::handleWeatherData()
{
    QNetworkReply* reply = qobject_cast<QNetworkReply*>(sender());
    if (!reply) return;

    QByteArray response = reply->readAll();
    reply->deleteLater();

    qDebug() << "[ë‚ ì”¨ ì‘ë‹µ JSON]" << QString::fromUtf8(response);

    QJsonParseError err;
    QJsonDocument doc = QJsonDocument::fromJson(response, &err);
    if (err.error != QJsonParseError::NoError) {
        qDebug() << "[JSON íŒŒì‹± ì—ëŸ¬]" << err.errorString();
        ui->label_weather->setText("ë‚ ì”¨ ì •ë³´ ì˜¤ë¥˜");
        return;
    }

    QJsonObject obj = doc.object();
    QJsonArray weatherArray = obj["weather"].toArray();
    if (weatherArray.isEmpty()) {
        qDebug() << "[ë‚ ì”¨ ì‘ë‹µ ì˜¤ë¥˜] weather ë°°ì—´ì´ ë¹„ì–´ ìˆìŒ";
        ui->label_weather->setText("ë‚ ì”¨ ì •ë³´ ì—†ìŒ");
        return;
    }

    QJsonObject weatherObj = weatherArray.first().toObject();
    QString description = weatherObj["description"].toString();
    QString iconCode = weatherObj["icon"].toString();
    double temp = obj["main"].toObject()["temp"].toDouble();

    ui->label_weather->setText(QString("%1 / %2Â°C").arg(description).arg(temp, 0, 'f', 1));

    // âœ… ì•„ì´ì½˜ ìš”ì²­ ë³„ë„ í•¨ìˆ˜ë¡œ ë¶„ë¦¬
    fetchWeatherIcon(iconCode);
}

void DetectMainWidget::handleWeatherIcon()
{
    QNetworkReply* reply = qobject_cast<QNetworkReply*>(sender());
    if (!reply) return;

    QByteArray imageData = reply->readAll();
    reply->deleteLater();

    QPixmap pix;
    if (!pix.loadFromData(imageData)) {
        qDebug() << "[ë‚ ì”¨ ì•„ì´ì½˜ ì˜¤ë¥˜] loadFromData ì‹¤íŒ¨";
        return;
    }

    ui->label_weatherIcon->setPixmap(pix.scaled(40, 40, Qt::KeepAspectRatio, Qt::SmoothTransformation));
}




